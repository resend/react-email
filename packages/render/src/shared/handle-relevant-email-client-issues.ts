import { decode } from 'he';

export const handleRelevantEmailClientIssues = (html: string): string => {
  return html
    .replaceAll(/href='([^']+)'/g, (match, content: string) => {
      const decodedContent = decode(content, {
        isAttributeValue: true,
      });
      console.log(decodedContent);
      if (decodedContent.includes("'")) {
        throw new Error(
          'Decoded href contains single quotes, which cannot be safely replaced without breaking the link. Keep your links free of duplicated quotes like this.',
        );
      }
      return match.replace(content, decodedContent);
    })
    .replaceAll(/href="([^"]+)"/g, (match, content: string) => {
      const decodedContent = decode(content, {
        isAttributeValue: true,
      });
      if (decodedContent.includes('"')) {
        throw new Error(
          'Decoded href contains double quotes, which cannot be safely replaced without breaking the href="...". Keep your links free of duplicated quotes like this.',
        );
      }
      return match.replace(content, decodedContent);
    });
};
