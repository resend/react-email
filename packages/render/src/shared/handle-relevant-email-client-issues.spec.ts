import { handleRelevantEmailClientIssues } from './handle-relevant-email-client-issues';

describe('handleRelevantEmailClientIssues', () => {
  // This is necessary because some click trackning mechanisms break when trying to the links with HTML entities with their own
  //
  // See https://github.com/resend/react-email/issues/1767#issuecomment-2747680180
  it('should decode href attributes with single quotes', () => {
    const input = `<a href='https://example.com/?utm_source=mailchimp&amp;utm_medium=email&amp;utm_campaign=example'>Click here</a><a href="https://example.com/?utm_source=mailchimp&amp;utm_medium=email&amp;utm_campaign=example">Click here</a>`;
    expect(handleRelevantEmailClientIssues(input)).toMatchSnapshot('');
  });
});
