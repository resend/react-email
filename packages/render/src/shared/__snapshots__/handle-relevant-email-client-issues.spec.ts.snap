// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleRelevantEmailClientIssues > should decode href attributes with single quotes 1`] = `"<a href='https://example.com/?utm_source=mailchimp&utm_medium=email&utm_campaign=example'>Click here</a><a href="https://example.com/?utm_source=mailchimp&utm_medium=email&utm_campaign=example">Click here</a>"`;
